### PR TITLE
Make e2e tests for new unprivileged functionality

### LINKED ISSUE #412: Need e2e tests for new unprivileged functionality
Add e2e tests for mounting SIF files unprivileged added in #386 and for unprivileged persistent overlay and writable tmpfs added in #411.

### PR DESCRIPTION
**Pull request by JasonYangShadow**
## Description of the Pull Request (PR):

Need e2e tests for new unprivileged functionality

### This fixes or addresses the following GitHub issues:

 - Fixes #412 

#### Before submitting a PR, make sure you have done the following:

- Read the [Guidelines for Contributing](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/apptainer/apptainer/blob/main/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md)
- Added tests to validate this PR, linted with `make check`  and tested this PR locally with a `make test`, and `make testall` if possible (see CONTRIBUTING.md).
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTORS.md)

### CODE REVIEW

**Comment by JasonYangShadow** [comment]
please merge this PR first
https://github.com/apptainer/apptainer/pull/607

**Comment by DrDaveD** [comment]
Please rebase this on the current main

File: CHANGELOG.md [hunk 1]
```diff
@@ -40,6 +40,7 @@ For older changes see the [archived Singularity change log](https://github.com/a
 - Add additional options to the build command for testing different fakeroot
   modes: `--userns` like the action flag and hidden options `--ignore-subuid`,
   `--ignore-fakeroot-command`, and `--ignore-userns`.
+- Add additional profile for testing unprivileged overlayfs driver.
```
Comment (DrDaveD) [inline CHANGELOG.md:43]: There's no need for a CHANGELOG entry here since only e2e tests are affected.

File: e2e/internal/e2e/profile.go [hunk 2]
```diff
@@ -37,6 +39,8 @@ var (
 	UserNamespaceProfile = Profiles[userNamespaceProfile]
 	// RootUserNamespaceProfile is the execution profile for root and a user namespace.
 	RootUserNamespaceProfile = Profiles[rootUserNamespaceProfile]
+	// OverlayProfile is the execution profile for unprivileged overlayfs mount
+	OverlayProfile = Profiles[overlayProfile]
```
Comment (DrDaveD) [inline e2e/internal/e2e/profile.go:43]: If this is only for unprivileged overlayfs let's put that in the name.  How about UserOverlayProfile.

File: internal/pkg/test/tool/require/require.go [hunk 3]
```diff
@@ -322,3 +322,11 @@ func MkfsExt3(t *testing.T) {
 		t.Skipf("mkfs.ext3 is too old and doesn't support -d")
 	}
 }
+
+// FuseOverlayfs check that whether fuse-overlayfs is available.
+func FuseOverlayfs(t *testing.T) {
```
Comment (DrDaveD) [inline internal/pkg/test/tool/require/require.go:327]: This function isn't needed  Use `bin.FindBin("fuse-overlayfs")` from "github.com/apptainer/apptainer/internal/pkg/util/bin"

File: e2e/internal/e2e/profile.go [hunk 4]
```diff
@@ -112,6 +116,16 @@ var Profiles = map[string]Profile{
...
 		optionForCommands: []string{"shell", "exec", "run", "test", "instance start"},
 	},
+	overlayProfile: {
+		name:              "Overlay",
+		privileged:        false,
+		hostUID:           origUID,
+		containerUID:      origUID,
+		defaultCwd:        "",
+		requirementsFn:    overlayRequirements,
+		apptainerOption:   "",
```
Comment (DrDaveD) [inline e2e/internal/e2e/profile.go:126]: There should be a `"--userns"` apptainerOption.

**Review by DrDaveD — CHANGES_REQUESTED** [review]
This is only one piece of #412.  There's still testing needed for using squashfuse and fuse2fs. It's fine to do this in multiple PRs but then don't say in the PR description that it fixes #412.

File: internal/pkg/test/tool/require/require.go (hunk 3 above)
Comment (DrDaveD) [reply internal/pkg/test/tool/require/require.go:327]: This also make me think that we didn't do anything similar for the fakeroot command.  Maybe there should be UserFakerootCmdProfile and FakerootCmdProfile (that's non-userns) that require that command.  They could probably be included in the PR that adds `--fakeroot` tests for action commands (but then use them also in the build `--fakeroot`.

File: e2e/internal/e2e/profile.go (hunk 4 above)
Comment (JasonYangShadow) [reply e2e/internal/e2e/profile.go:126]: > This is only one piece of #412. There's still testing needed for using squashfuse and fuse2fs. It's fine to do this in multiple PRs but then don't say in the PR description that it fixes #412.

sorry, my bad, after adding tests for overlay, I forgot the description of this issue. will add relative tests asap.

**Comment by JasonYangShadow** [comment]
> This is only one piece of #412. There's still testing needed for using squashfuse and fuse2fs. It's fine to do this in multiple PRs but then don't say in the PR description that it fixes #412.

sorry, my bad, forgot to add tests for squashfuse and fuse2fs after adding tests for overlay. will add those tests asap.

File: e2e/internal/e2e/profile.go (hunk 4 above)
Comment (JasonYangShadow) [reply e2e/internal/e2e/profile.go:126]: Not sure about the details, but I thought we did not need an unprivileged user namespace to use fuse-overlayfs?

File: CHANGELOG.md (hunk 1 above)
Comment (JasonYangShadow) [reply CHANGELOG.md:43]: got it!

**Comment by JasonYangShadow** [comment]
@DrDaveD 
I squashed previous commits.
but it looks like some tests failed. It looks like this test fails due to some changes in the remote image.
I can also reproduce this fatal locally using the main branch. 
```
2022-08-17T08:44:31.6957260Z === CONT  TestE2E/PAR/ENV/environment_manipulation
2022-08-17T08:44:31.6958142Z     apptainercmd.go:703: 
2022-08-17T08:44:31.6959307Z         "/usr/local/bin/apptainer exec docker://ghcr.io/apptainer/lolcow /bin/sh -c echo $PATH" output:
2022-08-17T08:44:31.6960221Z         INFO:    Converting OCI blobs to SIF format
2022-08-17T08:44:31.6960836Z         INFO:    Starting build...
2022-08-17T08:44:31.6961307Z         Getting image source signatures
2022-08-17T08:44:31.6961983Z         Copying blob sha256:7709a11df2eb89c55eb4cb1b08fb622921adba3ac21d600e659c2fab07e32338
2022-08-17T08:44:31.6962787Z         Copying config sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a
2022-08-17T08:44:31.6963396Z         Writing manifest to image destination
2022-08-17T08:44:31.6963774Z         Storing signatures
2022-08-17T08:44:31.6965552Z         FATAL:   Unable to handle docker://ghcr.io/apptainer/lolcow uri: while building SIF from layers: conveyor failed to get: unsupported image-specific operation on artifact with type "application/vnd.unknown.config.v1+json"
2022-08-17T08:44:31.6966147Z         
2022-08-17T08:44:31.6966537Z     apptainercmd.go:703: got 255 as exit code and was expecting 0: exit status 255
2022-08-17T08:44:31.6967291Z         waiting for command "/usr/local/bin/apptainer exec docker://ghcr.io/apptainer/lolcow /bin/sh -c echo $PATH"
2022-08-17T08:44:31.6967870Z         github.com/apptainer/apptainer/e2e/internal/e2e.TestEnv.RunApptainer.func1
2022-08-17T08:44:31.6968435Z         	github.com/apptainer/apptainer/e2e/internal/e2e/apptainercmd.go:678
2022-08-17T08:44:31.6968958Z         github.com/apptainer/apptainer/e2e/internal/e2e.TestEnv.RunApptainer.func2
2022-08-17T08:44:31.6969575Z         	github.com/apptainer/apptainer/e2e/internal/e2e/apptainercmd.go:703
2022-08-17T08:44:31.6969890Z         runtime.goexit
```

File: e2e/run/run.go [hunk 5]
```diff
@@ -206,5 +306,8 @@ func E2ETests(env e2e.TestEnv) testhelper.Tests {
 		"inaccessible home":    c.issue409,
 		"passphrase encrypted": c.testRunPassphraseEncrypted,
 		"PEM encrypted":        c.testRunPEMEncrypted,
+		"user overlayfs":       c.testUserOverlayfs,
+		"squashfuse mount":     c.testSquashfuseMount,
+		"ext3 mount":           c.testExt3Mount,
```
Comment (DrDaveD) [inline e2e/run/run.go:311]: I think the tests are good, but let's be more consistent in the naming.  I suggest
```
"fuse overlayfs":         c.testFuseOverlayfs,
"fuse squash mount": c.testFuseSquashMount,
"fuse ext3 mount":     c.testFuseExt3Mount,
```

**Review by DrDaveD — CHANGES_REQUESTED** [review]
The commits still need a bit more squashing, as there are 4 now instead of 1.  The description of the commit should be what's changing, not the fact that it is a squashing of commits.

File: e2e/run/run.go (hunk 5 above)
Comment (JasonYangShadow) [reply e2e/run/run.go]: resolved and squashed all commits into one
https://github.com/apptainer/apptainer/pull/615/commits/017cb82de2a3167b83e98eeb19e156cc52939e7c

**Review by DrDaveD — APPROVED** [review]

**Pull request merged**